### PR TITLE
Redirect TFile::Open() calls to XRootD protocol if possible

### PR DESCRIPTION
**Pull request by amadio**
In a meeting in IT I saw that ~200PB/year are being read from the `eoshome` instance via fuse, and I thought that maybe ROOT at least could try to do something smart and just redirect fuse traffic to XRootD if possible. The patch in this pull request does just that. When I talked to Axel about making this work, we wanted the redirection to not happen when using `file://file.root`, but it created some inconsistencies when using `TFile::Open` vs using `TChain` so the final implementation always attempts to redirect.

Below is a demo session of ROOT opening a file on EOS with (a former version of) the patch:
```sh
$ pwd
/eos/home-a/amadio
$ ls *.root
Run2012BC_DoubleMuParked_Muons.root
$ root.exe
   ------------------------------------------------------------------
  | Welcome to ROOT 6.27/01                        https://root.cern |
  | (c) 1995-2022, The ROOT Team; conception: R. Brun, F. Rademakers |
  | Built for linuxx8664gcc on Oct 28 2022, 09:49:15                 |
  | From heads/redirect-xrootd@v6-09-01-24773-gd85df4c5e9            |
  | With c++ (GCC) 8.5.0 20210514 (Red Hat 8.5.0-15)                 |
  | Try '.help'/'.?', '.demo', '.license', '.credits', '.quit'/'.q'  |
   ------------------------------------------------------------------

root [0] auto f1 = TFile::Open("file://Run2012BC_DoubleMuParked_Muons.root");
root [1] auto f2 = TFile::Open("Run2012BC_DoubleMuParked_Muons.root");
root [2] f1->GetName()
(const char *) "Run2012BC_DoubleMuParked_Muons.root"
root [3] f2->GetName()
(const char *) "root://eoshome-a.cern.ch//eos/user/a/amadio/Run2012BC_DoubleMuParked_Muons.root"
root [4] .q
$
```
and with a slightly modified `df102_NanoAODDimuonAnalysis` tutorial to avoid `TChain`:
```diff
void df102_NanoAODDimuonAnalysis(const char* filename)
 {
    // Enable multi-threading
    ROOT::EnableImplicitMT();
 
-   // std::cout << "Using filename: " << filename << std::endl;
+   auto f = TFile::Open(filename);
+   auto t = f->Get<TTree>("Events");
 
-   ROOT::RDataFrame df("Events", filename);
+   ROOT::RDataFrame df(*t);
```

I got the following:

```sh
$ time ./dimuon file://Run2012BC_DoubleMuParked_Muons.root 
Info in <TCanvas::Print>: pdf file dimuon_spectrum.pdf has been created
Events with exactly two muons: pass=31104343   all=61540413   -- eff=50.54 % cumulative eff=50.54 %
Muons with opposite charge: pass=24067843   all=31104343   -- eff=77.38 % cumulative eff=39.11 %
13.34
$ time ./dimuon Run2012BC_DoubleMuParked_Muons.root 
Info in <TCanvas::Print>: pdf file dimuon_spectrum.pdf has been created
Events with exactly two muons: pass=31104343   all=61540413   -- eff=50.54 % cumulative eff=50.54 %
Muons with opposite charge: pass=24067843   all=31104343   -- eff=77.38 % cumulative eff=39.11 %
8.81
$
```

**Note:** With the final version of the patch, instead of using `file://`, which now does not disable the redirection anymore, one should add `TFile.CrossProtocolRedirects: no` to their `rootrc` configuration file in order to disable this feature (I checked that it works).

### CODE REVIEW

File: io/io/src/TFile.cxx
Comment (hageboeck) [inline io/io/src/TFile.cxx:4043]: Does this work?
```suggestion
            expandedUrl = xurl;
```

Comment (amadio) [reply io/io/src/TFile.cxx:4043]: Remnant from when I had this change in `TUrl`, will fix in next push.

**Comment by eguiraud** [comment]
> it just so happens that the file:// gets stripped if you create an RDataFrame("MyTree", "file://file.root") and then the redirection always happens

Guilherme tells me that it's actually TChain that strips the `file://` from the filename. @pcanal could it not do that? :smile:

**Comment by amadio** [comment]
Please clang-format as well...

**Comment by amadio** [comment]
> Guilherme tells me that it's actually TChain that strips the file:// from the filename. @pcanal could it not do that? smile

The first call to `TFile::Open` happens here:
```
#0  0x00007ffff580b7c0 in TFile::Open(char const*, char const*, char const*, int, int)@plt ()
   from libTree.so
#1  0x00007ffff5856d3a in TChain::LoadTree (this=0x1027b50, entry=0) at core/base/inc/TString.h:244
#2  0x00007ffff58545d2 in TChain::GetListOfBranches (this=0x1027b50) at tree/tree/src/TChain.cxx:1116
#3  0x00007ffff3ee03a6 in (anonymous namespace)::GetBranchNamesImpl (t=..., bNamesReg=std::set with 0 elements, 
    bNames=std::vector of length 0, capacity 0, analysedTrees=std::set with 1 element = {...}, friendName="", 
    allowDuplicates=true) at tree/dataframe/src/RLoopManager.cxx:142
#4  0x00007ffff3ee0e7c in ROOT::Internal::RDF::GetBranchNames[abi:cxx11](TTree&, bool) (t=..., 
    allowDuplicates=<optimized out>) at tree/dataframe/src/RLoopManager.cxx:341
#5  0x00007ffff3ee0f1d in ROOT::Detail::RDF::RLoopManager::GetBranchNames[abi:cxx11]() (this=0x16502a0)
    at tree/dataframe/src/RLoopManager.cxx:1015
#6  0x000000000046121e in ROOT::RDF::RInterface<ROOT::Detail::RDF::RLoopManager, void>::Filter(std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >) ()
#7  0x000000000045e376 in df102_NanoAODDimuonAnalysis(char const*) ()
#8  0x000000000045eb60 in main ()
```
by this time, the `file://` prefix is already stripped.

**Comment by eguiraud** [comment]
The `file://` prefix is lost here:

https://github.com/root-project/root/blob/8cc176c9ee7dde3b72501a2b657534bc6bd10418/tree/tree/src/TChain.cxx#L485

(this is in `TChain::AddFile` but `TChain::Add` also ends up here). `basename` is the file name stripped of the `file://` prefix and it is what's used to form the file name that's passed to the `TChainElement` constructor later (which is what stores the information about each sub-tree in the chain).

**Comment by amadio** [comment]
Thanks for the detailed analysis, @eguiraud! I will see what happens when I don't strip the prefix. Should be fine, since this goes into `TFile::Open` but better get confirmation from @pcanal as well.

**Comment by amadio** [comment]
Removing the special casing leads to this:
```
execve("./df", ["./df", "Run2012BC_DoubleMuParked_Muons.r"...], 0x7ffd482bdf58 /* 49 vars */) = 0
getxattr("file:Run2012BC_DoubleMuParked_Muons.root", "eos.url.xroot", NULL, 0) = -1 ENOENT (No such file or directory)
access("/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", F_OK) = 0
access("/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", R_OK) = 0
openat(AT_FDCWD, "/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", O_RDONLY) = 4
...
```
So that breaks the redirection, since now `file:` gets added somewhere... :disappointed:

**Comment by amadio** [comment]
Bummer, `xurl.data()` is [`const` until C++14](https://en.cppreference.com/w/cpp/string/basic_string/data), so replaced with `&xurl[0]`.

**Comment by amadio** [comment]
Dropped the draft status, as this should be now safe to merge.

**Review by hageboeck — COMMENTED** [review]
It looks like this will work, but I suggest to add info to the docstring and possibly to the release notes.

File: io/io/src/TFile.cxx
Comment (hageboeck) [inline io/io/src/TFile.cxx:79]: Still needed?

Comment (amadio) [reply io/io/src/TFile.cxx:79]: I removed this on CentOS and it still compiled, I will try to remove, let's see. I only re-indented because clang-format complained.

Comment (eguiraud) [reply io/io/src/TFile.cxx:79]: unrelated to this PR though?

File: io/io/src/TFile.cxx
Comment (eguiraud) [inline io/io/src/TFile.cxx:4041]: should probably add a comment here mentioning that `getxattr` only works with plain paths and if `url` starts with `file://` (or any protocol for what that matters, if I understand correctly) `getxattr` will fail and we'll skip the rest of the logic.

**Review by eguiraud — APPROVED** [review]
LGTM, although we should wait for Philippe's or Axel's approval this in order to merge this.

I don't know whether this deserves an info message when the redirection happens, at least for the first release in which we do it?

And of course it's not optimal that `file://path/to/file` removes the redirect for plain TFile::Opens but it doesn't work with TChain (and therefore RDF) -- but I would rather redirect _more_ than less :smile: in 99.9% of cases that's the right thing to do.

File: io/io/src/TFile.cxx
Comment (hageboeck) [reply io/io/src/TFile.cxx:79]: I overlooked the `-` part of the diff. I really thought that the include had been added here so yes, it's unrelated.

File: io/io/src/TFile.cxx
Comment (amadio) [reply io/io/src/TFile.cxx:4041]: Yes, I guess the comment I added is not enough. I will update it with your suggestion.

File: README/ReleaseNotes/v628/index.md
Comment (hageboeck) [inline README/ReleaseNotes/v628/index.md:91]: Here is an attempt to explain a bit more to the user why they want this feature:
```suggestion
- Faster reading from EOS.
  A new cross-protocol redirection has been added to `TFile::Open` to allow files on EOS
  mounts to be opened via the XRootD protocol. This considerably improves read speeds.
  The redirection uses the `eos.url.xroot` extended file attribute that is present on files
  in EOS. The attribute can be viewed with `getfattr -n eos.url.xroot [file]` on the command
  line. When the URL passed into `TFile::Open` is a plain file name for a file on an EOS mount,
  the extended attribute is used to attempt the redirection to the XRootD protocol. If the URL
  starts with `file://`, the redirection is disabled. Note that when ROOT files are opened via
  `TChain`, the `file://` prefix gets stripped from the URL automatically. Therefore, the
  redirection will always be attempted in a `TChain`.
  If a redirection fails, the file is opened using the plain file path as before. All redirections
  can be disabled setting `TFile.CrossProtocolRedirects` to 0 in `rootrc`.
```

**Comment by hageboeck** [comment]
FYI: A 2GB opendata file in eos home, read & processed using 4 threads with RDF.
| Throughput vs time | Aggregated Throughput |
| --- | --- |
| image | image |

File: io/io/src/TFile.cxx
Comment (pcanal) [inline io/io/src/TFile.cxx:4039]: This should be moved after the expansion of the filename (line 4057), so that environment variable and others things are already substituted.

**Review by pcanal — CHANGES_REQUESTED** [review]
It seems the code needs to be moved a tad bit later.

I don't understand if the choice of making the feature optional on whether `file://` is there is a technical convenience (in which case we could strip it using TUrl for example or something else) or a interface design choice (the doc seems to indicate the earlier).

**Comment by pcanal** [comment]
This looks great (delta the code location and the interface question).  Thanks!

**Comment by amadio** [comment]
Thanks, @pcanal. I will move the code. Hopefully after that we can merge. I also had a request to test this “across the ocean”, so if you could try to compare before/after by running something simple using data at CERN (or let me know of a file publicly avaiable via XRootD in Fermilab or other location in the US), I would appreciate it.

**Comment by amadio** [comment]
As for the behavior when using `file://`, I guess we can follow what other redirections are doing. I don't have a strong feeling, but having `file://` disable the redirection was indeed the intention, apart from the fact that with `TChain` it cannot be disabled due to the fact that it strips the prefix. If you think we should be stripping in `TUrl`, I could add that, then the redirections would be controlled only by `TFile.CrossProtocolRedirects` in the configuration (which is enabled by default).

**Comment by pcanal** [comment]
On `file:\\` I hesitate on whether this is a user request to use the posix interface no matter what or whether the performance difference between `posix-eos` and `xrootd-eos` is always so great that we should make the right choice for the user.

**Comment by amadio** [comment]
@Axel-Naumann, maybe you can comment as well? What should we do? Strip `file://` and redirect or let `file://` disable the redirection?

**Comment by pcanal** [comment]
> so if you could try to compare before/after by running something simple using data at CERN

CMS currently does not mount CERN's eos through fuse at FNAL and does not export (without grid certificate) any data.

**Comment by Axel-Naumann** [comment]
If `file://` cannot consistently prevent redirection (also for `TChain`) then this mechanism is inconsistent, the behavior will be hard to understand / predict, and we should always redirect (or fix the `TChain` case).

I find it a good idea to have an `Info` once per process that the URL is redirected.

**Comment by amadio** [comment]
This seems to be the only other place where the configuration setting is checked:
https://github.com/root-project/root/blob/master/io/io/src/TFile.cxx#L4201-L4209

I think people may be annoyed by the `Info` messages and we'd need some static storage to store if the message has been emitted or not for the current process, so I'm not a fan of adding it (although I have no strong feelings about this either).

I will add stripping of `file://` prefix so that the behavior is consistent (always redirect) then. I tried and didn't find an easy way to get `file://` to disable the redirection in the case of `TChain`.

**Comment by amadio** [comment]
@pcanal, I guess you mean line 4057 after this patch, so [this line](https://github.com/root-project/root/blob/6858acbeec33a4230a097d04e82c5fe8102b3b25/io/io/src/TFile.cxx#L4057), right? I will move the code and update this pull request.

**Comment by amadio** [comment]
@Axel-Naumann I've updated the patch to use `TUrl` and always redirect. I updated the code comment and release notes to match. The way to disable the redirection then will be to set `TFile.CrossProtocolRedirects` to `0` in `rootrc`. Now the behavior should be consistent between plain `TFile::Open` calls and `TChain`.

@pcanal I tried moving the code after line 4057 and using `expandedUrl` but it started showing errors like this when using `TChain`:
```
Error in <TNetXNGFile::Open>: [ERROR] Server responded with an error: [3001] Required argument not present
```

So I moved it back where it was before and use only the input `url` to `TFile::Open` as is. I tested that this now works both for plain `TFile::Open` calls as well as with `TChain` and the redirection always happens as long as the file being opened is on EOS. The problem was that the `exapandedUrl` sometimes contained only the base EOS management URL, which broke it.

**Comment by pcanal** [comment]
> The problem was that the exapandedUrl sometimes contained only the base EOS management URL, which broke it.

Can you give a concrete illustration?

File: io/io/src/TFile.cxx
Comment (amadio) [reply io/io/src/TFile.cxx:4039]: This did not work as mentioned in another comment. Resolving.

File: io/io/src/TFile.cxx
Comment (amadio) [reply io/io/src/TFile.cxx]: I didn't see a zombie being returned, even when I tested without proper credentials, but I added a check just in case. When you can read via `/eos` but not via `root://`, a lot of error messages about credentials appear in the output:
```
Error in <TNetXNGFile::Open>: [ERROR] Server responded with an error: [3010] Unable to give access - user access restricted - unauthorized identity used ; Permission denied
```
However, in the end you get a `nullptr` from the attempted `TFile::Open` and fallback to opening the file via FUSE.

**Comment by amadio** [comment]
> > The problem was that the exapandedUrl sometimes contained only the base EOS management URL, which broke it.
> 
> Can you give a concrete illustration?

Here is a session with `strace` with the redirection disabled:
```
$ strace ./dimuon Run2012BC_DoubleMuParked_Muons.root 2>&1 | grep Muons
execve("./dimuon", ["./dimuon", "Run2012BC_DoubleMuParked_Muons.r"...], 0x7ffd2e39de38 /* 49 vars */) = 0
access("/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", F_OK) = 0
access("/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", R_OK) = 0
openat(AT_FDCWD, "/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", O_RDONLY) = 4
lstat("/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", {st_mode=S_IFREG|0644, st_size=2244449133, ...}) = 0
lstat("/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", {st_mode=S_IFREG|0644, st_size=2244449133, ...}) = 0
access("/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", F_OK) = 0
access("/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", R_OK) = 0
openat(AT_FDCWD, "/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", O_RDONLY) = 5
lstat("/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", {st_mode=S_IFREG|0644, st_size=2244449133, ...}) = 0
lstat("/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", {st_mode=S_IFREG|0644, st_size=2244449133, ...}) = 0
access("/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", F_OK) = 0
access("/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", R_OK) = 0
openat(AT_FDCWD, "/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", O_RDONLY) = 5
lstat("/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", {st_mode=S_IFREG|0644, st_size=2244449133, ...}) = 0
lstat("/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", {st_mode=S_IFREG|0644, st_size=2244449133, ...}) = 0
write(1, "Muons with opposite charge: pass"..., 97Muons with opposite charge: pass=24067843   all=31104343   -- eff=77.38 % cumulative eff=39.11 %
```
one with the redirection enabled:
```
$ strace ./dimuon Run2012BC_DoubleMuParked_Muons.root 2>&1 | grep Muons
execve("./dimuon", ["./dimuon", "Run2012BC_DoubleMuParked_Muons.r"...], 0x7fff35f90878 /* 49 vars */) = 0
getxattr("Run2012BC_DoubleMuParked_Muons.root", "eos.url.xroot", NULL, 0) = 79
getxattr("Run2012BC_DoubleMuParked_Muons.root", "eos.url.xroot", "root://eoshome-a.cern.ch//eos/us"..., 79) = 79
write(1, "Muons with opposite charge: pass"..., 97Muons with opposite charge: pass=24067843   all=31104343   -- eff=77.38 % cumulative eff=39.11 %
```
and finally, one with the code moved to the place you suggested (using `expandedUrl`):
```bash
$ strace ./dimuon Run2012BC_DoubleMuParked_Muons.root 2>&1 | grep Muons
execve("./dimuon", ["./dimuon", "Run2012BC_DoubleMuParked_Muons.r"...], 0x7ffccf746158 /* 49 vars */) = 0
getxattr("Run2012BC_DoubleMuParked_Muons.root", "eos.url.xroot", NULL, 0) = 25
getxattr("Run2012BC_DoubleMuParked_Muons.root", "eos.url.xroot", "root://eoshome-a.cern.ch/", 25) = 25
access("/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", F_OK) = 0
access("/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", R_OK) = 0
openat(AT_FDCWD, "/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", O_RDONLY) = 12
lstat("/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", {st_mode=S_IFREG|0644, st_size=2244449133, ...}) = 0
lstat("/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root", {st_mode=S_IFREG|0644, st_size=2244449133, ...}) = 0
getxattr("Run2012BC_DoubleMuParked_Muons.root", "eos.url.xroot", NULL, 0) = 79
getxattr("Run2012BC_DoubleMuParked_Muons.root", "eos.url.xroot", "root://eoshome-a.cern.ch//eos/us"..., 79) = 79
getxattr("Run2012BC_DoubleMuParked_Muons.root", "eos.url.xroot", NULL, 0) = 79
getxattr("Run2012BC_DoubleMuParked_Muons.root", "eos.url.xroot", "root://eoshome-a.cern.ch//eos/us"..., 79) = 79
write(1, "Muons with opposite charge: pass"..., 97Muons with opposite charge: pass=24067843   all=31104343   -- eff=77.38 % cumulative eff=39.11 %
```
It is a bit puzzling (doesn't always happen), but as you can see, in the case using `expandedUrl`, the URL returned is only the base management URL, and the file in the end is opened via FUSE once. This doesn't always happen, but I thought I'd be on the safe side and avoid potential problems.

File: README/ReleaseNotes/v628/index.md
Comment (amadio) [reply README/ReleaseNotes/v628/index.md]: I took some of the suggestions in the end, as I had to touch the code and go through the CI again anyway.

**Comment by amadio** [comment]
More puzzling, here are two consecutive runs without changing anything (using `expandedUrl`):
```bash
lcgapp-centos8-physical amadio $ cat test.C
#include "TFile.h"

void test(const char* filename, const char* options)
{
  auto f = TFile::Open(filename, options);
}

int main(int argc, char **argv)
{
   test(argv[1], argv[2]);
}
lcgapp-centos8-physical amadio $ cat ~/.rootrc && strace ./test test.root 2>&1 | grep "test.root"
TFile.CrossProtocolRedirects: yes
execve("./test", ["./test", "test.root"], 0x7fffd10a02b8 /* 49 vars */) = 0
getxattr("test.root", "eos.url.xroot", NULL, 0) = 54
getxattr("test.root", "eos.url.xroot", "root://eoshome-a.cern.ch//eos/us"..., 54) = 54
access("/eos/home-a/amadio/test.root", F_OK) = 0
access("/eos/home-a/amadio/test.root", R_OK) = 0
openat(AT_FDCWD, "/eos/home-a/amadio/test.root", O_RDONLY) = 11
lstat("/eos/home-a/amadio/test.root", {st_mode=S_IFREG|0644, st_size=2244449133, ...}) = 0
lstat("/eos/home-a/amadio/test.root", {st_mode=S_IFREG|0644, st_size=2244449133, ...}) = 0
lcgapp-centos8-physical amadio $ cat ~/.rootrc && strace ./test test.root 2>&1 | grep "test.root"
TFile.CrossProtocolRedirects: yes
execve("./test", ["./test", "test.root"], 0x7fff277b6488 /* 49 vars */) = 0
getxattr("test.root", "eos.url.xroot", NULL, 0) = 53
getxattr("test.root", "eos.url.xroot", "root://eoshome-a.cern.ch//eos/us"..., 53) = 53
```

**Comment by amadio** [comment]
For the record, the build will fail because some tests based on reference files now have some unexpected warning:
```
+ld: warning: -undefined dynamic_lookup may not work with chained fixups
+ld: warning: -undefined dynamic_lookup may not work with chained fixups
```
These are not caused by my changes, so I will merge this if those are the only failures.

**Comment by Axel-Naumann** [comment]
Yup sorry I'm on it.

**Comment by pcanal** [comment]
> This doesn't always happen, ...

This actually quite concerning ... In the use case you show, there should be not practical difference as the filename should be strictly the same before and after the expand.  (As opposed to load "/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root" vs "$MY_EOS_DIR/Run2012BC_DoubleMuParked_Muons.root" where the redirection would work only for the 1st one without the expansion).

In the first strace, I see:

```
getxattr("Run2012BC_DoubleMuParked_Muons.root", "eos.url.xroot", NULL, 0) = 79
getxattr("Run2012BC_DoubleMuParked_Muons.root", "eos.url.xroot", "root://eoshome-a.cern.ch//eos/us"..., 79) = 79
```
vs
```
getxattr("Run2012BC_DoubleMuParked_Muons.root", "eos.url.xroot", NULL, 0) = 25
getxattr("Run2012BC_DoubleMuParked_Muons.root", "eos.url.xroot", "root://eoshome-a.cern.ch/", 25) = 25
```
which "seems" to mean that goven the exact same input `getxattr` returns different result.  Why would that be?

>More puzzling, here are two consecutive runs without changing anything (using expandedUrl):

That seems "consistent" with the "inconsistent" behavior I just talked about.  So there is something wrong about `expandfilename` that we need to fix since:

(a) it is semantically necessary to call it
(b) it behaves oddly in this case but also probably also in other case without us knowing.

so since we have a running failing example of the odd behavior we may as well investigate it now.

**Comment by amadio** [comment]
> > This doesn't always happen, ...
> 
> This actually quite concerning ... In the use case you show, there should be not practical difference as the filename should be strictly the same before and after the expand. (As opposed to load "/eos/home-a/amadio/Run2012BC_DoubleMuParked_Muons.root" vs "$MY_EOS_DIR/Run2012BC_DoubleMuParked_Muons.root" where the redirection would work only for the 1st one without the expansion).
> 
> In the first strace, I see:
> 
> ```
> getxattr("Run2012BC_DoubleMuParked_Muons.root", "eos.url.xroot", NULL, 0) = 79
> getxattr("Run2012BC_DoubleMuParked_Muons.root", "eos.url.xroot", "root://eoshome-a.cern.ch//eos/us"..., 79) = 79
> ```
> 
> vs
> 
> ```
> getxattr("Run2012BC_DoubleMuParked_Muons.root", "eos.url.xroot", NULL, 0) = 25
> getxattr("Run2012BC_DoubleMuParked_Muons.root", "eos.url.xroot", "root://eoshome-a.cern.ch/", 25) = 25
> ```
> 
> which "seems" to mean that given the exact same input `getxattr` returns different result. Why would that be?

Good question, and I'm not sure. This seems to be a problem with EOS (see below). @apeters1971 Do you know what could cause this? To reproduce, all you need is to copy a file with a different name, then call `getfattr` on it and it will return the wrong URL, at least the first time (but potentially several times):

```bash
$ cp hsimple.root mytest.root
$ getfattr -n eos.url.xroot hsimple.root && getfattr -n eos.url.xroot mytest.root
# file: hsimple.root
eos.url.xroot="root://eoshome-a.cern.ch//eos/user/a/amadio/hsimple.root"

# file: mytest.root
eos.url.xroot="root://eoshome-a.cern.ch/"
```

So I guess we *could* move the code after the expansion after all, as the problem is not coming from ROOT, but from EOS. When the wrong URL is returned, all that happens is that the opening via XRootD fails and it falls back to use FUSE, so I guess it should still be fine to merge this, if only for allowing people to test more easily once ROOT master is installed somewhere in CVMFS.

**Review by pcanal — APPROVED** [review]
Thanks.

**Comment by amadio** [comment]
Ah! Perfect timing, Philippe! Thanks! Merging.

**Pull request merged**
